### ISSUE TITLE
Make Cross-Platform

### ISSUE THREAD

**Issue by shanazzlowe** [JosefNemec/Playnite#59]
For those that have Linux or Mac, and not Windows.

**Comment by JosefNemec** [comment]
Not planned in near future, see #38 for more details.

**Comment by JosefNemec** [comment]
I will leave this open for other people to see. Also it's not definite NO, it's just not possible at the moment.

**Comment by JosefNemec** [comment]
Note to myself: check out [NoesisGUI](http://www.noesisengine.com/) for possible multi-platform interface.

**Comment by aaronfranke** [comment]
+1 Linux/Mac cross-platform. I recommend using Mono instead of .NET.

**Comment by johannesgunnar** [comment]
Would love to see a Linux version of this! <3

**Comment by LucLarabie** [comment]
Would also want a Linux version, hope it will be a thing someday.

**Comment by jharmer95** [comment]
Might want to look into .NET Core 3.0 (will come out sometime early next year), it supports WPF
![image](https://user-images.githubusercontent.com/21198728/46448138-5cec4080-c753-11e8-9aa0-185f14760060.png)

**Comment by JosefNemec** [comment]
.Net Core 3.0 will not make WPF multiplatform, it only means that you will be able to implement WPF applications using Core instead of Framework. But those applications will only run on Windows.

**Comment by JosefNemec** [comment]
Possible good news for Playnite cross platform. Microsoft just [open sourced](https://blogs.msdn.microsoft.com/dotnet/2018/12/04/announcing-net-core-3-preview-1-and-open-sourcing-windows-desktop-frameworks/) WPF, so there is a hope if somebody ports it to other platforms :)

**Comment by murlakatamenka** [comment]
@JosefNemec came here to share this news but clearly I'm late. Glad to see you're aware of it.

**Comment by trymeouteh** [comment]
This would be great. I want to make my old gaming pc a linux gaming rig and playnite is great!

**Comment by jarrard** [comment]
+1 For Linux, having a alternative to Lutris would be nice, however I know it will be a challenge because people will expect you to integrate wine container support, and possible custom wine build downloads like Lutris. Would be welcome non the less.

**Comment by TBK** [comment]
nix support should be "relatively easy" with .NET Core 3.0 - https://github.com/dotnet/wpf#status

**Comment by JosefNemec** [comment]
@TBK *nix support is literally impossible right now. WPF being open source doesn't mean that it runs on Linux, it's still Windows only and based on Windows only technologies (DirectX). Somebody would need to port first, which is definitely not "relatively easy".

**Comment by DanielGilbert** [comment]
@JosefNemec Highly appreciate your effort in supporting Linux. You might want to take a look at https://github.com/AvaloniaUI/Avalonia for a cross-platform, WPF-like UI. Although it's Beta, it's a pretty advanced and stable Beta. Just as an idea.

Might not solve the CefSharp issue, though.

Keep up the great work!

**Comment by Ruedii** [comment]
I can chart a few steps to making cross platform.

I browsed over the code and specific WPF code is limited as is other Windows specific code, but it is scattered through the codebase.

I am not sure if YAML is used embedded in HTML.  If so that might be difficult to work around.  Moving to pure HTML5/CSS would be the only option for working around this, but this opens the door for other features, such as remote browsing of your library from any web browser.

As of other parts they should be able to be isolated.  Only a handful of windows-specific functions are anywhere that isn't already platform specific.   Most of those could be isolated to small subsection written specifically for each platform.

Process tracking needs to be done per-platform.
Registry functions would need to be replaced with non-windows alternatives.
Most game launching plugins would each have substantial per-platform segments.

**Comment by aaronfranke** [comment]
By the way, Microsoft has announced that .NET Framework is being discontinued, and that after .NET Core 3.1 there will be one .NET called .NET 5: https://devblogs.microsoft.com/dotnet/introducing-net-5/

So, moving to Mono or .NET Core would not only help make Playnite cross-platform, but it would also prevent Playnite from being left behind using an unsupported framework.

Microsoft is also trying to make Mono and .NET Core function as drop-in replacements for each other, and Mono is designed to re-implement .NET Framework, so it should be fairly easy to convert Playnite to Mono and then optionally to .NET Core.

**Comment by JosefNemec** [comment]
This is what has to happen for me to consider doing Linux/macOS. And this is for me personally for the official Playnite releases that I develop a provide some support for.

1. There's a big enough demand for it in the first place.

I don't want to end up maintaining these versions for just handful of people. Mainly because there is quire similar alternative for Linux called Lutris, which seems to be quite popular.

2. WPF is ported to Linux/mac or Avalonia reaches mature state with functionality on par with WPF.

At this point reworking UI to something completely different is not an option for me. I like working with XAML based UIs and completely reworking UI stack is not something I want to do (especially that that there are no good alternatives with good .NET support).

3. CefSharp is ported to Linux/macOS or some good alternative is found.

CefSharp is quite critical part of Playnite for 3rd party integrations and we would not be able to port them over without it. As an alternative we would have to find some usable alternative, which I'm not currently aware of (CefGlue probably being the only one, but it's not as good as CefSharp from what I saw in past).

**Comment by sferra** [comment]
That sounds reasonable. 2) and 3) seem to be external factors which you don't/won't influence directly.
But how do we make 1) measurable and how would you define "big enough"? : )

**Comment by JosefNemec** [comment]
I don't know how big is big enough. I will evaluate that properly once #2 and #3 are sorted out :)

**Comment by slavikme** [comment]
Please, help me understand something.
If this is just a launcher, why it should be developed with .NET?
There are tons of other (and good) frameworks that have cross-platform support, like Electron, Have and more.

**Comment by JosefNemec** [comment]
> Please, help me understand something.
> If this is just a launcher, why it should be developed with .NET?
> There are tons of other (and good) frameworks that have cross-platform support, like Electron, Haxe and more.

Because I like working with .NET (and C#) and I don't like developing desktop apps using web technologies.

**Comment by slavikme** [comment]
Thanks for the clarification.

**Comment by Scaledish** [comment]
It is really a killer, I love this launcher, but I need to find one that will work on both my linux and windows pc

**Comment by JosefNemec** [comment]
I'm just going to leave [this](https://ccifra.github.io/PortingWPFAppsToLinux/Overview.html) here to make sure I don't forget about it in case I decide to do Linux version in future.

**Comment by MicahMosier** [comment]
Hello Josef, this is an outstanding project and I appreciate the effort you have put into this. Have you thought more on porting this to linux? Or maybe have a timeline?

**Comment by JosefNemec** [comment]
Before somebody mentions that we should use [this](https://devblogs.microsoft.com/dotnet/introducing-net-multi-platform-app-ui/), new .NET Multi-platform App UI announced yesterday, here are my thoughts about it:

- It's based on Xamarin.Forms, which is pretty bad on desktop (current version that is). Very limited compared to WPF and forcing you to use design patters suitable mainly for mobile platforms.
- Linux will not be officially supported by Microsoft and will by maintained by the community, which makes it dead on arrival for Linux.
- Looks like they are removing theming support.
- It's in very early stage of development, so it will take years before it reaches level of maturity to be usable for us.

**Comment by bverkron** [comment]
I see little mention of WINE. Has anyone tried running this via WINE?

**Comment by shakeyourbunny** [comment]
I don't see much sense into this, because you have to run the games also in the same WINE context as PlayNite would be in.

**Comment by djibux** [comment]
Well @bverkron, it runs OK with Wine provided that you have mono installed (package mono-complete on Debian) and use Winetricks to install corefonts and .Net (I went for version 4.8 but 4.6.2 should be enough).
The app has some graphical glitches but it's very usable.
As a library manager Playnite is really awesome.  I don't use it to launch games but use Lutris for that.  It creates a different wine prefix for each game which it nice when you run multiple games that require different tweaks.
I would love a native version :)
Hope that helps.

**Comment by shakeyourbunny** [comment]
Any news on this? :)

**Comment by JosefNemec** [comment]
No news, don't expect Linux version any time soon.

**Comment by trymeouteh** [comment]
It would be cool if someone could rewrite Playnite into a cross platform code base that is open source like Qt or something. I don't expect the JosefNemec to start over again but if there is anyone who is good at rewriting apps into new code bases that could rewrite Playnite and if JosefNemec will be willing to abandon the current code base and continue develop of Playnite in the new code base from there on then then Linux and Mac support will become a reality.

**Comment by Vundult** [comment]
Playnite is the only thing I truly miss from windows. Really do hope it gets ported someday. I loved using it to manage all my games, and it's so much better than Lutris. Lutris UI is ugly, and hardly stands up against playnite, but it's the only real option right now. Closest thing I've found is Gamehub for Linux, but that's still lacking tools needed to run games via WINE.

**Comment by Bluscream** [comment]
Since i recently switched from Windows to Manjaro i would love to see an official way to run this on here.

**Comment by zaroth** [comment]
With Valve announcement to release [Steam Deck](https://store.steampowered.com/steamdeck), Linux gaming should see a surge in compatibility (as Valve already announced) and popularity. Just found this project (Playnite) today, it seems cool and I'd love to try this out, but Windows-only is a non-starter for me, sadly. Leaving this comment here to be notified of potential developments, though.

**Comment by frostworx** [comment]
Totally forgot about this project, so thanks for the bump @zaroth!
Did anybody test recently playnite wine/proton compatibility?
might be an interesting feature to add into my launcher then

**Comment by JosefNemec** [comment]
Please do not report any issues related to Playnite running on unsupported platforms. And please use forums/reddit if you want to discuss unofficial ways of running Playnite on Linux. Keep this issue only for talk about official support.

As far as official support, I'm not looking into it right now and running via Wine/Proton will have ton of issues since we need access to native system to properly support all features and integrations with other Linux apps. I'm only considering doing native version, I will personally not work any emulated version.

**Comment by shakeyourbunny** [comment]
I'd second a **native** linux port, Playnite.

Avalonia ( https://github.com/AvaloniaUI/Avalonia ) has matured in the meantime quite a bit.

Perhaps you would setup an extra Patreon goal or do collect some funding for a port to Linux? You could count me in on funding this, but it would also need some sort of announcement. Informing news sites to want to make a port with funding would help. Linux people do want to fund projects if the goal is reasonably, so publishing this to GamingOnLinux, Phoronix, some other more gaming oriented news sites would help. In addition this also would raise the awareness of the project in general.

**Comment by JosefNemec** [comment]
Avalonia is not mature enough for project like Playnite, I'm following it closely and it's currently not usable for Playnite.

**Comment by boehs** [comment]
> Avalonia is not mature enough for project like Playnite, I'm following it closely and it's currently not usable for Playnite.

What work needs to be done?

**Comment by JosefNemec** [comment]
There's still plenty of performance, rendering, usability and stability issues. Even the API is still not frozen for 1.0 release.

**Comment by OsoaGitHub** [comment]
i feel a playnite port of linux would be welcomed now more than ever due to the recent gaming on linux news

**Comment by qd7sa** [comment]
👀

**Comment by dragonDScript** [comment]
Given that Steam Deck is a thing, would you rewrite the entire thing in Avalonia? Avalonia is much more mature now.

**Comment by kwilliams1987** [comment]
> Given that Steam Deck is a thing, would you rewrite the entire thing in Avalonia? Avalonia is much more mature now. There's UNO platform too.

Would make far more sense to convert from .NET Framework to .NET 6.0/MAUI for UX if he ever decided to port it.

No need to port the entire system and would make Windows/macOS/Linux support much simpler.

The biggest hurdle for cross compatibility probably isn't even porting the UI, it will be rewriting all the interop code with the various third-party launchers for the other platforms.

**Comment by JosefNemec** [comment]
The biggest hurdle **IS** porting the UI. Read my previous messages on the topic. Btw MAUI is a no go, read my reasoning in previous messages.

**Comment by woj-tek** [comment]
>Btw MAUI is a no go, read my reasoning in previous messages. 

If anyone is wondering and doesn't want to read whole topic (and searching for "MAUI" doesn't point to the message because it's not mentioned there explicitly) it's this message: https://github.com/JosefNemec/Playnite/issues/59#issuecomment-631246016

**Comment by ThisNekoGuy** [comment]
> Not planned in near future, see #38 for more details.

@JosefNemec 
Why couldn't UNO Platform be used? It's meant for keeping compatibility for the Windows API and make applications cross-platform.
Stuff like the native Windows calculator has been ported to Linux using it before.

**Comment by JosefNemec** [comment]
From some research I did like a year ago, Avalonia looked like only realistic candidate. I don't remember what I didn't like about UNO, I think they primarily "just" wrap native platforms controls, which is no good for app like Playnite with high level of customizability and theming. Also performance was no that great from what I remember.

But I will look into it again when times comes (if it comes) to start working on Linux port.

**Comment by arghanath007** [comment]
Still only for windows?

**Comment by shakeyourbunny** [comment]
Any news on this? 

A port to Steam Deck / Linux would be the killer...

**Comment by zocker-160** [comment]
_I know this is OT, I can delete it again, when this kind of comment is not welcome here_

I have been monitoring this issue here for quite some time now and I wonder how big the "need" / interest actually is.

Back when SteamDesk was announced I worked on a controller-first UI based on MiniGalaxy using Qt but stopped due to lack of interest.
If there is actually demand for that, I would be willing to continue development again (would probably take around one month of work to get into a usable state).
This my current state:
![Screenshot_20220330_212556](https://user-images.githubusercontent.com/36563538/160914972-4359ccc7-3d8f-4927-9065-cf4b924135e7.png)

**Comment by JosefNemec** [comment]
Current plan is to port non-UI code to .NET 6/7 this year and swap UI stack for something different after that (most likely Avalonia, which has Linux support), then we'll see.

If you want to discuss different projects (like mentioned MiniGalaxy), please use different place. From my point of view there's nothing more to discuss until ports I mentioned are finished.

**Comment by DillFrescott** [comment]
I know its not a definitive yes, but it would be hugely awesome to be able to enjoy Playnite on Linux! By the way thank you for this program!

**Comment by DocMAX** [comment]
Linux version please, thanks.

**Comment by JosefNemec** [comment]
2025 update

Upcoming major release (P11) will be on modern open source and multi-platform .NET runtime but will still use Window only WPF based UI. This is for several reasons, one of the main ones being that Avalonia's developer experience is not, in my opinion, good enough to be used in project like Playnite. Playnite heavily depends on community made extensions and themes, often developed by hobbits developers, not professionals, and Avalonia is simply not user friendly as much as I would like it to be for us.

We are 100% moving to Avalonia at some point in future, it's just not going to be with Playnite 11 release.

So current plan is to release Playnite 11, continue developing it for Windows only and once Avalonia's DX improves (they are promising major tools improvements with their upcoming Accelerate program), I will start porting P11 to Avalonia.

Once we have polished Windows version of Avalonia based Playnite, I'll start working on Linux version. I don't want to make just straight port of Windows version, that wouldn't be good enough. I would really like to have Linux specific feature available on release, like Wine/Proton specific features, Epic and GOG plugins using Legendary and heroic-gogdl respectively, stuff like that.

To sum this up, Linux version is definitely something I plan to work on in future, it's just going to happen later than originally planned.

As far as macOS version, no idea about that. It's definitely not something I would look into sooner than having Linux version more or less done.

**Comment by JosefNemec** [comment]
2025 update 2

I'm planning to move to Linux in 2026 after P11 is done, and since Playnite is my personal blocker for the move, I will try to make some Linux version in 2026. Definitely not fully featured version on parity with Windows version of P11, and probably desktop mode only, but something daily drivable with Linux specific features (Wine/Proton integration for example).

If that goes well, I'll start looking into P11 Avalonia porting proper say I mentioned in previous update.
